src/api/ucc.h

```diff
@@ -0,0 +1,709 @@
+/**
+ * @file ucc.h
+ * @date 2020
+ * @copyright Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCC_H_
+#define UCC_H_
+
+#include <api/ucc_def.h>
+#include <api/ucc_version.h>
+#include <api/ucc_status.h>
+#include <stdio.h>
+#include <stdbool.h>
+
+/**
+ * *************************************************************
+ *                   Library initialization and finalize 
+ * *************************************************************
+ */
+
+typedef enum {
+    UCC_THREAD_MODE       = UCC_BIT(0),
+    UCC_COLL_TYPES        = UCC_BIT(1),
+    UCC_REDUCTION_TYPES   = UCC_BIT(2),
+    UCC_SYNC_TYPE         = UCC_BIT(3) 
+} ucc_lib_params_mask_t;
+
+typedef enum {
+    UCC_OP_USERDEFINED      = UCC_BIT(0),
+    UCC_OP_SUM              = UCC_BIT(1),
+    UCC_OP_PROD             = UCC_BIT(2),
+    UCC_OP_MAX              = UCC_BIT(3),
+    UCC_OP_MIN              = UCC_BIT(4),
+    UCC_OP_AND              = UCC_BIT(5),
+    UCC_OP_OR               = UCC_BIT(6),
+    UCC_OP_XOR              = UCC_BIT(7),
+    UCC_OP_LAND             = UCC_BIT(8),
+    UCC_OP_LOR              = UCC_BIT(9),
+    UCC_OP_LXOR             = UCC_BIT(10),
+    UCC_OP_BAND             = UCC_BIT(11),
+    UCC_OP_BOR              = UCC_BIT(12),
+    UCC_OP_BXOR             = UCC_BIT(13),
+    UCC_OP_MAXLOC           = UCC_BIT(14),
+    UCC_OP_MINLOC           = UCC_BIT(15)
+} ucc_reduction_op_t;
```

Comment:
IIRC we wanted this to be a union of enum and pointer, to support passing a user-defined operation.

Reply:
Aug 19th WG Discussion: 
* A separate field for “user-defined” reductions will be added. 
* User need to pass UCC_OP_USERDEFINED option and reduction function to UCC